Initial RISC-V jit based on IR

unknownbrackets opened the pull request:
~This isn't quite ready, as it's still slower than interpreter (mostly during loading though, probably needs load/store implemented.)~

This is now often faster than the interpreter, although not so much when floats/vecs are heavily used.

#17771, #17773, and #17776 should be merged first since this enables jit by default.

Need to test more games, but so far most run pretty well (barring some IR bugs.)  Tests also pass (except 2 that also fail for IR.)

Flush to zero looks like it will be a problem (and I recall Gods Eater Burst requires it for save validation...)  RISC-V doesn't seem to intend to expose it.

Notes on the implementation:
 * Currently based on IR heavily, I wanted to see the pros/cons of using its block cache which is pretty different (also gives me preload which may help me find bugs easier.)  Probably makes sense to switch to a "full fledged" backend using the regular block cache.
 * A decent number of ALU related ops implemented now, and does have static alloc etc.  No floats yet.
 * Crash handler is not well tested (but seems to work.)
 * Rounding hasn't been thought through much.
 * Periodic trampolines or constant polls may make sense for far jumps or references to C++ globals.
 * SCRATCH1/SCRATCH2 should probably not be a0/a1, worried it'll cause confusion in func calls (save/load regs etc.)  But they are regs that can be used in 16-bit instructions...

-[Unknown]

unknownbrackets committed:
Debugger: Handle missing crash/block ptrs better.

unknownbrackets committed:
riscv: Initial staffolding for IR based jit.

unknownbrackets committed:
riscv: Initial untested dispatcher.

The minimum to actually, probably, running code.  Pretty slow.

unknownbrackets committed:
irjit: Define a specific IRReg type.

unknownbrackets committed:
riscv: Add a register cache for jit.

Not yet actually used.  Might be buggy.

unknownbrackets committed:
riscv: Add jit for some initial instructions.

unknownbrackets committed:
riscv: Add some safety to pointerifying.

We have to clear the upper bits in case of sign extension or other things.

unknownbrackets committed:
riscv: Implement some simple assign instructions.

hrydgard commented:
Very cool to try IR -> native! Will look at this soon, and looking forward to following this. 

What do you mean by periodic trampolines?

unknownbrackets commented:
RISC-V can only jump +/- 1MB, so even going back to the dispatcher, rounding, save/load regs, block linking, etc. requires a register and jump.  It does have AUIPC (which the emitter is currently not optimally using) which allows a +/-31 bit jump in two instructions, so maybe that's enough for semi-local jumps.

But anywhere we're calling (or loading) things outside the jit space, constructing a 64-bit immediate can take 6 instructions (with a second temporary reg) and for example clang uses literal pools for 64-bit constants instead.  So a periodic trampoline/pool might make sense, or might not - if only from a codesize perspective.

-[Unknown]

hrydgard commented:
We might just want to make more effort to squeeze everything together into a 31-bit range then, so we can use that 2-instruction combo. Not sure how feasible though.

unknownbrackets commented:
Well, non-jit code (by which I mean, Advance, syscalls, coreState, etc.) won't always be there.  It's similiar to the problem on x86 with RIP addressing.

-[Unknown]

hrydgard commented:
Well - they aren't in range currently, but they could be made to be, with careful enough usage of mmap etc. We could suballocate all our executable buffers from one larger one for example and even have that suballocated from a larger block that we manually put stuff like coreState in, though it will be some extra complexity.

unknownbrackets committed:
riscv: Stub out more IR compilation categories.

unknownbrackets committed:
riscv: Implement load and store ops.

unknownbrackets committed:
riscv: Avoid needless save/load around compile.

unknownbrackets committed:
riscv: Implement the simplest exits.

unknownbrackets committed:
riscv: Allow dirty pointers, explicitly.

unknownbrackets committed:
riscv: Reduce jit codesize a bit.

unknownbrackets committed:
riscv: Use R_RA as a temporary for calls.

This is the most logical thing, since we're about to write it anyway.

unknownbrackets committed:
riscv: Correct depointerify on FlushAll().

unknownbrackets committed:
riscv: Implement conditional exits.

unknownbrackets committed:
riscv: Try to keep regs normalized, track.

Since we can't address the lower 32-bits only in compares, this can help
us avoid renormalizing before a compare.

unknownbrackets committed:
riscv: Implement logic ops.

unknownbrackets committed:
riscv: Implement shifts and compares.

unknownbrackets committed:
riscv: Implement conditional moves.

unknownbrackets committed:
riscv: Fix block lookup issues.

unknownbrackets committed:
riscv: Fix min/max normalization.

unknownbrackets committed:
riscv: Optimize SetConst a bit.

unknownbrackets committed:
riscv: Enable jit by default.

unknownbrackets committed:
riscv: Use direct SLI/SLIU instructions.

Derp, I forgot these existed on RISC-V for a moment.

unknownbrackets committed:
riscv: Implement multiply instructions.

unknownbrackets committed:
riscv: Cleanup some pointerification flags.

hrydgard commented on Core/MIPS/RiscV/RiscVRegCache.cpp:
```diff
@@ -0,0 +1,1075 @@
+	ar[reg].isDirty = (mapFlags & MIPSMap::DIRTY) == MIPSMap::DIRTY;
+	if ((mapFlags & MIPSMap::NOINIT) != MIPSMap::NOINIT) {
+		if (mipsReg == MIPS_REG_ZERO) {
+			// If we get a request to load the zero register, at least we won't spend
```
in the comment: s/load/map

hrydgard reviewed (approved):
Again, really good to see a first stab at a IR->native JIT, thanks for doing this!

I had somehow hoped for the post-IR regcache to get simpler, turns out it's not really, for good reasons. At least there's less of a requirement for the instruction emulations to collapse immediates since that's already done by the IR optimizer, which is nice.

Overall, this looks great. Just some nit-picking and some comments that don't need addressing.

I'm surprised at the lack of denormal flushing support in the ISA, that's gonna be an issue for things like Proton and other emulators too...

unknownbrackets commented on Core/MIPS/RiscV/RiscVJit.cpp:
```diff
@@ -0,0 +1,557 @@
+	*rhs = NormalizeR(inst.src2, allowOverlap ? 0 : inst.dest, rhsTempReg);
+}
+
+RiscVReg RiscVJit::NormalizeR(IRRegIndex rs, IRRegIndex rd, RiscVReg tempReg) {
```
We can't.  There's no ANDW, and there's also no BGEW or SLTIW.  Those don't exist, so sometimes you need to normalize.

-[Unknown]

hrydgard replied:
Ah, I didn't realize. I guess that's the end of that idea, then.

unknownbrackets commented on Core/MIPS/RiscV/RiscVJit.cpp:
```diff
@@ -0,0 +1,557 @@
+		}
+	}
+
+	// Note: a properly constructed block should never get here.
```
Well, the loop can get here just fine, it's just that jit execution shouldn't.  I could QuickCallFunction to a log/assert.

-[Unknown]

hrydgard replied:
Oh, right. Well, might not be worth the trouble.

unknownbrackets commented on Core/MIPS/RiscV/RiscVJit.cpp:
```diff
@@ -0,0 +1,557 @@
+
+void RiscVJit::CompIR_Generic(IRInst inst) {
+	// For now, we're gonna do it the slow and ugly way.
+	// Maybe there's a smarter way to fallback?
```
I'll make it check the exit flag on IRMeta to at least skip the return value check.

-[Unknown]
